keras/utils/data_utils.py

```diff
@@ -612,26 +612,49 @@ def __init__(self, generator,
                  seed=None):
         self.wait_time = wait_time
         self._generator = generator
-        self._use_multiprocessing = use_multiprocessing
+        if os.name is 'nt' and use_multiprocessing is True:
+            # On Windows, avoid **SYSTEMATIC** error in `multiprocessing`:
+            # `TypeError: can't pickle generator objects`
+            # => Suggest multithreading instead of multiprocessing on Windows
+            raise ValueError('Using a generator with `use_multiprocessing=True`'
+                             ' is not supported on Windows (no marshalling of'
+                             ' generators across process boundaries). Instead,'
+                             ' use single thread/process or multithreading.')
+        else:
+            self._use_multiprocessing = use_multiprocessing
         self._threads = []
         self._stop_event = None
         self._manager = None
         self.queue = None
         self.seed = seed
 
-    def start(self, workers=1, max_queue_size=10):
-        """Kicks off threads which add data from the generator into the queue.
-
-        # Arguments
-            workers: number of worker threads
-            max_queue_size: queue size
-                (when full, threads could block on `put()`)
-        """
-
-        def data_generator_task():
+    def data_generator_task(self):
+        if self._use_multiprocessing is False:
+            while not self._stop_event.is_set():
+                with self.genlock:
+                    try:
+                        if self.queue is not None and self.queue.qsize() < self.max_queue_size:
+                            # On all OSes, avoid **SYSTEMATIC** error in multithreading mode:
+                            # `ValueError: generator already executing`
+                            # => Serialize calls to infinite iterator/generator's next() function
+                            generator_output = next(self._generator)
+                            self.queue.put((True, generator_output))
+                        else:
+                            time.sleep(self.wait_time)
```

Comment:
Sleeping with a lock held seems bad. Shouldn't it be sufficient to guard this line with the lock:
```
generator_output = next(self._generator)
```

Reply:
In my (admittedly limited) experience with generators, the thread is mostly waiting on getting an available slot in the queue rather than on getting samples from the generator. I guess this is all very dependent on whatever your use case scenario is.
With the above in mind, I believe it is much better sleeping than busy waiting on the queue not being full anymore (that's essentially what the self.queue.qsize() < self.max_queue_size test is about).
Or course, a much better fix would have the lock replaced with a condition variable and signalling with the queue as wells as the generator. I assumed you decided not to go that way in the first place because you wanted to make it easy on developers to write generators. Or, perhaps, because of the limitations of multithreading in Python, you decided there was absolutely no point on making it much harder on developers to have to adhere to a much more sophisticated synchronization mechanism for a feature that is so clobbered by the gil anyway...

Reply:
Well, it's not really a busy wait since its mostly sleeping. However, I have to agree that there isn't that much to be gained by holding the lock shorter.

I think it would only matter if putting the object in the queue is a time consuming action (which it isn't for the threading case).

Reply:
I was also bothered by the sleep in the lock and thought to correct it. Even though the code can be refactored to avoid it, WE CANT lock only the `generator_output = next(self._generator)`. This will not work. I thought to put a comment here, if someone tries to do this on the future.

This PR makes the queue to have fixed size. If we don't protect both the iterator and the addition to the queue, we can end up with threads being stack while trying to add to the queue. Because the stop() operation joins and waits for the threads to finish, this will not happen. To avoid sleeping in a lock, the function needs to be heavily refactored and to echo @philferriere might not be 100% worth it.

Reply:
In the future, I'll just remove this use case. I suspect having `workers > 1 and use_multiprocessing == False` doesn't gives any real speedup. 

I'll do some profiling and post my findings here.

Also, we should try to mimic the Ordered Enqueuer so this class will get heavily refactored anytime soon anyway. (By the end of August)

Reply:
I spotted this PR because I was trying to debug a starvation effect on a Multi-GPU setup. Even though I understand the reasons why @philferriere patched it, the introduced lock has a nasty side-effect on speed. Effectively this turns the method into single-thread, as the most heavy component is covered in lock.

This PR tries to protect against situations where the Generator is not thread-safe. Nevertheless if it is thread-safe, removing the lock gets you a **4-5 times speed improvement**. You can actually get a significant speed up even if `workers > 1 and use_multiprocessing == False`. This is because numpy and other C libraries release the GIL in many functions. Indeed, Python does not have a true multi-threading mechanism, but C libraries can help you achieve better parallelism. This is especially true in Computer Vision apps, where heavy computations happen in the Generators (data augmentation) and careful locking can give massive benefits. 

Finally I understand that Sequences might be a better option than Generators when it comes to multi-threading. Nevertheless they require "random-access" functionalities that might not be always available. Personally I think that some parts of this PR (mainly the locks) should not have been kept. The lock causes the most heavy computations to run on single threat. This also means that if your Generator is thread-safe you can't benefit from it. IMO if your generator is not thread-safe you should not use `workers>1`.

@Dref360 @fchollet Originally I considered building a case for removing the locks, provide benchmarks and send a PR. It seems that you have specific plans for this class that are going the opposite direction. Would you be open discussing this through?

Reply:
With use_multiprocessing=False, how do you avoid `ValueError : generator already executing` without adding a Lock and therefore lose all speedup?

Reply:
Let me provide some context and then move to benchmarks. Here are the key Keras releases for this discussion:
2.0.5: All Generator classes for Images used Iterator logic.
2.0.6: Sequences introduced.
2.1.3: Locks introduced in GeneratorEnqueuer

Before the introduction of Sequences (<=2.0.5), all Generator classes in Keras implemented `next()`. They were also thread-safe and as a result the `GeneratorEnqueuer` did not require any locks. On the other hand, python generator methods did not work well as they were throwing `ValueError : generator already executing`. This was fixed by @philferriere in this PR and it was released on 2.1.3.

When I talk about thread-safe generators, I'm not refering to python generator methods but to Generator classes like those used by Keras [before 2.0.5](https://github.com/keras-team/keras/blob/d4b618bf23d83b1a7306bc6816bd509c23bba15d/keras/preprocessing/image.py#L682). These classes used an Iterator and made sure by using minimal locking that all the heavy computations are done ourside the lock. Since C libs can release the GIL, this gave speed improvements. When Sequences were introduced, many of the classes were rewritten and instead of `next()` they implemented `__getitem__()`. Sequences are better when concurrency is necessary, nevertheless not all use cases can implement a "random-access" logic; still they can be thread-safe. Thus having the ability to build thread-safe generator classes is still useful IMO. The lock introduced on 2.1.3, stops those valid cases from using proper parallelism.

Below I provide a snippet that can be ran with and without the lock (just comment out line [650](https://github.com/keras-team/keras/blob/master/keras/utils/data_utils.py#L650)). To produce something more than a toy example, I use the `ImageDataGenerator` class with realistic configuration to load real images (note that I am aware that ImageDataGenerator will return a Sequence but that has no effect here; for an Iterator-based implementation you can take the older version of the class from Keras 2.0.5 and get the same results. What we need here is a thread-safe "iterator" with heavy computations in non-locking parts.). 

Snippet:
```python
import time
from keras.preprocessing import image
from keras.utils.data_utils import GeneratorEnqueuer

it = image.ImageDataGenerator().flow_from_directory('/path/to/images', target_size=(224, 224), batch_size=512)

reader = GeneratorEnqueuer(it, use_multiprocessing=False)
reader.start(workers=16, max_queue_size=16)
g = reader.get()

start = time.time()
n = 100
for i in range(n):
	x=g.next()

total_time = time.time()-start
print('Total time %f sec; Average time: %f sec' % (total_time, total_time/n))

```

**WITH LOCK:** Total time 161.514681 sec; Average time: 1.615147 sec
**WITHOUT LOCK:** Total time 32.234057 sec; Average time: 0.322341 sec

That's 5 times faster, provided that the underlying iterator is thread-safe. As you can see, `use_multiprocessing=False and workers>1` can be quite useful because many C libraries release the GIL. Unfortunately the lock introduced above, even though admittedly fixes the case of Python generators, it makes it impossible for thread-safe iterators to work fast. Note that this benchmark does not conflict yours. If I run the same code without the Enqueuer, its time is the same as with the Enqueuer & the lock (exactly what you said earlier).

In my opinion, removing the lock makes sense (or maybe checking the instance type and/or the existence of specific methods to understand if the lock is necessary?), because if you are using non-thread-safe Python generators you should not use `workers>1`. If you use thread-safe iterators though, it absolutely makes sense to use parallelism especially if you implement them with minimal locking. Finally I want to mention that this is not a theoretical discussion. I started investigating this problem because we observed GPU starvation and 4-5 times slower execution times on real-world applications after Keras 2.1.3. Our problem is solved by removing the lock and making a couple of minor changes on the queue configuration.

If that's interesting to you, I can draft a PR so that you can check it out and decide if it's something you want to use.

Reply:
Please draft a PR. I think we should keep a UX-friendly way of handling python generators. (ie. a good clear message stating that they should set workers=1)

Reply:
Sounds good. I'll put something together in the weekend and I'll tag you to get your feedback.

Reply:
> In my opinion, removing the lock makes sense (or maybe checking the instance type and/or the existence of specific methods to understand if the lock is necessary?), because if you are using non-thread-safe Python generators you should not use workers>1.

Yeah, I think it's a good idea to add some specific member variable to indicate that a generator is safe for running multi-threaded so it can be detected at runtime. That could then trigger a warning and enable the locking to serialize the `next()` calls with `workers > 1`.